### ISSUE TITLE
Source domain isn't used, and using localhost:2368 instead

### ISSUE THREAD

**Issue by imtomjorge** [Fried-Chicken/ghost-static-site-generator#65]
Ghost version: 5
NPM Version: 6.14.16
Node Version: v12.22.12
Ghost Static Site Generator version: v1.1.4
OS distro + version: Ubuntu 20.04.4 LTS

Description: When I try to launch gssg, it seems that the flag "--sourceDomain" is not used. 
Instead, it performs multiple wget commands on http://localhost:2368. 

Logs: 
```
root@my-server:~# gssg --sourceDomain http://myghosthosted-onanotherserver.mydomain.org/ --productionDomain http://static.mydomain.org/ --dest myblog-static-folder
Fetching: http://localhost:2368/
ERROR: null
Using Command: wget -q --show-progress --recursive --timestamping --page-requisites --no-parent --no-host-directories --restrict-file-name=unix --directory-prefix myblog-static-folder --content-on-error  --trust-server-names http://localhost:2368/
Fetching: http://localhost:2368/sitemap.xsl
ERROR: null
Using Command: wget -q --show-progress --recursive --timestamping --page-requisites --no-parent --no-host-directories --restrict-file-name=unix --directory-prefix myblog-static-folder --content-on-error  --trust-server-names http://localhost:2368/sitemap.xsl
Fetching: http://localhost:2368/sitemap.xml
ERROR: null
Using Command: wget -q --show-progress --recursive --timestamping --page-requisites --no-parent --no-host-directories --restrict-file-name=unix --directory-prefix myblog-static-folder --content-on-error  --trust-server-names http://localhost:2368/sitemap.xml
myblog-static-folder/sitemap.xml http://localhost:2368/sitemap.xml Error: ENOENT: no such file or directory, open '/root/myblog-static-folder/sitemap.xml'
    at Object.openSync (fs.js:462:3)
    at Object.readFileSync (fs.js:364:35)
    at fetchUrlHelper (/usr/lib/node_modules/ghost-static-site-generator/src/helpers/fetchUrlHelper/fetchUrlHelper.js:47:31)
    at Array.forEach (<anonymous>)
    at generateStaticSite (/usr/lib/node_modules/ghost-static-site-generator/src/commands/generateStaticSite.js:35:10) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/root/myblog-static-folder/sitemap.xml'
}
Fetching: http://localhost:2368/404
ERROR: null
Using Command: wget -q --show-progress --recursive --timestamping --page-requisites --no-parent --no-host-directories --restrict-file-name=unix --directory-prefix myblog-static-folder --content-on-error  --trust-server-names http://localhost:2368/404
Fetching: http://localhost:2368/public/ghost.css
ERROR: null
Using Command: wget -q --show-progress --recursive --timestamping --page-requisites --no-parent --no-host-directories --restrict-file-name=unix --directory-prefix myblog-static-folder --content-on-error  --trust-server-names http://localhost:2368/public/ghost.css
Fetching: http://localhost:2368/public/ghost.min.css
ERROR: null
Using Command: wget -q --show-progress --recursive --timestamping --page-requisites --no-parent --no-host-directories --restrict-file-name=unix --directory-prefix myblog-static-folder --content-on-error  --trust-server-names http://localhost:2368/public/ghost.min.css
Fetching: http://localhost:2368/public/404-ghost.png
ERROR: null
Using Command: wget -q --show-progress --recursive --timestamping --page-requisites --no-parent --no-host-directories --restrict-file-name=unix --directory-prefix myblog-static-folder --content-on-error  --trust-server-names http://localhost:2368/public/404-ghost.png
Fetching: http://localhost:2368/public/404-ghost@2x.png
ERROR: null
Using Command: wget -q --show-progress --recursive --timestamping --page-requisites --no-parent --no-host-directories --restrict-file-name=unix --directory-prefix myblog-static-folder --content-on-error  --trust-server-names http://localhost:2368/public/404-ghost@2x.png
Error: ENOENT: no such file or directory, copyfile '/root/myblog-static-folder/404/index.html' -> '/root/myblog-static-folder/404.html'
    at Object.copyFileSync (fs.js:1907:3)
    at copy404PageHelper (/usr/lib/node_modules/ghost-static-site-generator/src/helpers/copy404PageHelper/copy404PageHelper.js:15:8)
    at generateStaticSite (/usr/lib/node_modules/ghost-static-site-generator/src/commands/generateStaticSite.js:40:5) {
  errno: -2,
  syscall: 'copyfile',
  code: 'ENOENT',
  path: '/root/myblog-static-folder/404/index.html',
  dest: '/root/myblog-static-folder/404.html'
}
Domain: http://localhost:2368
Static site generated at: /root/myblog-static-folder
Site generated in: 229.252ms
```

**Comment by imtomjorge** [comment]
The issue also persists on Debian Buster:

root@ghost-static:~# lsb_release -d
Description:	Debian GNU/Linux 11 (bullseye)
root@ghost-static:~# gssg --version
1.1.4

**Comment by mewmix** [comment]
Same issue on WSL 2 ! No issues on my Mac tho?

**Comment by xalisys** [comment]
any update, i am facing same issue on fedora?

**Comment by rmarchant** [comment]
As a workound, I just manually update the `/usr/lib/node_modules/ghost-static-site-generator/src/constants/OPTIONS.js` file, and replace the lines that reference `localhost:2368` with the desired hostname:port.

**Comment by ItsIgnacioPortal** [comment]
Another workaround is to modify your `/etc/hosts` file

**Comment by dmd** [comment]
> Another workaround is to modify your `/etc/hosts` file

Now I'm just a humble chicken farmer but I'd suggest that having `localhost` point anywhere other than `localhost` is a supremely bad idea and likely to break all *sorts* of things.

**Comment by ItsIgnacioPortal** [comment]
> > Another workaround is to modify your `/etc/hosts` file
> 
> Now I'm just a humble chicken farmer but I'd suggest that having `localhost` point anywhere other than `localhost` is a supremely bad idea and likely to break all *sorts* of things.

Obviously you shouldn't keep this modification forever. Just modify it before running gssg, and restore it afterwards hahah

**Comment by wegotoeleven** [comment]
+1, same issue on Alpine

**Comment by Nikoos** [comment]
I had the same issue as you, however after doing some research on Internet, I was able to run a static site generation with the following command:

`gssg --domain "https://localhost:8080" --url "http://192.168.57.221:8081" --dest docs`

I found this command here : [https://dev.to/eli_xciv/from-ghost-blog-to-static-site-pke](https://dev.to/eli_xciv/from-ghost-blog-to-static-site-pke).

Nikos

**Comment by wunderhund** [comment]
Seeing a similar issue when running in a node:14 docker container. When I run Nikos' command above and point gssg to a ghost host other than localhost, certain files are still requested from localhost. In this case the container is connected to a docker network where it can reach ghost at http://ghost:2368:
```
root@700458ead844:/app# gssg --domain "http://ghost:2368" --url "http://blog.craigabutler.com" --ignore-absolute-paths
Fetching: http://ghost:2368/
index.html                            100%[======================================================================>]  32.98K  --.-KB/s    in 0.005s
robots.txt                            100%[======================================================================>]     107  --.-KB/s    in 0s
assets/built/screen.css?v=2124e23fcd  100%[======================================================================>]  41.63K  --.-KB/s    in 0.005s
public/cards.min.js?v=2124e23fcd      100%[======================================================================>]   7.23K  --.-KB/s    in 0s
public/cards.min.css?v=2124e23fcd     100%[======================================================================>]  25.48K  --.-KB/s    in 0.004s
test-post/index.html                  100%[======================================================================>]  13.74K  --.-KB/s    in 0.001s

[ ... ]

Fetching: http://ghost:2368/sitemap.xsl
sitemap.xsl                           100%[======================================================================>]   6.39K  --.-KB/s    in 0s
Fetching: http://ghost:2368/sitemap.xml
sitemap.xml                           100%[======================================================================>]     638  --.-KB/s    in 0s
Fetching: http://localhost:2368/sitemap-pages.xml
ERROR: null
Using Command: wget -q --show-progress --recursive --timestamping --page-requisites --no-parent --no-host-directories --restrict-file-name=unix --directory-prefix static --content-on-error  --trust-server-names http://localhost:2368/sitemap-pages.xml
statichttp://localhost:2368/sitemap-pages.xml http://localhost:2368/sitemap-pages.xml Error: ENOENT: no such file or directory, open '/app/statichttp:/localhost:2368/sitemap-pages.xml'
    at Object.openSync (fs.js:498:3)
    at Object.readFileSync (fs.js:394:35)
    at fetchUrlHelper (/usr/local/lib/node_modules/ghost-static-site-generator/src/helpers/fetchUrlHelper/fetchUrlHelper.js:47:31)
    at Array.forEach (<anonymous>)
    at /usr/local/lib/node_modules/ghost-static-site-generator/src/helpers/fetchUrlHelper/fetchUrlHelper.js:56:11
    at Parser.<anonymous> (/usr/local/lib/node_modules/ghost-static-site-generator/node_modules/xml2js/lib/parser.js:304:18)
    at Parser.emit (events.js:400:28)
    at SAXParser.onclosetag (/usr/local/lib/node_modules/ghost-static-site-generator/node_modules/xml2js/lib/parser.js:262:26)
    at emit (/usr/local/lib/node_modules/ghost-static-site-generator/node_modules/sax/lib/sax.js:624:35)
    at emitNode (/usr/local/lib/node_modules/ghost-static-site-generator/node_modules/sax/lib/sax.js:629:5) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/app/statichttp:/localhost:2368/sitemap-pages.xml'
}
```

The affected files are:

- sitemap-pages.xml
- sitemap-posts.xml
- sitemap-authors.xml
- sitemap-tags.xml

I haven't figured out what the problem is, but the fact that they're all XML files makes me think that it's some kind of issue in [replaceXmlUrlsHelper](https://github.com/Fried-Chicken/ghost-static-site-generator/tree/master/src/helpers/replaceXmlUrlsHelper) where `OPTIONS.SOURCE_DOMAIN_WITHOUT_PROTOCOL` is loading the default 'localhost' value instead of whatever was passed into the `--domain` option.

To address @imtomjorge's original post, looking at the commit history, I see that the `--sourceDomain` option was added in [2d8bfd8](https://github.com/Fried-Chicken/ghost-static-site-generator/commit/2d8bfd8ddd85846d14f9ab44e59d827e70cc151e), after the 1.1.4 release. In fact, GitHub thinks that the current version in the repository is 1.2.0, but the version provided by npm is still 1.1.4. @Fried-Chicken is there any reason you haven't pushed the newer version to npm?

**Comment by Nikoos** [comment]
@wunderhund you're correct, I remember using npm instead of sources from this repo, sorry for the trouble.

**Comment by ItsIgnacioPortal** [comment]
> Seeing a similar issue when running in a node:14 docker container. When I run Nikos' command above and point gssg to a ghost host other than localhost, certain files are still requested from localhost. In this case the container is connected to a docker network where it can reach ghost at http://ghost:2368:
> 
> ```
> root@700458ead844:/app# gssg --domain "http://ghost:2368" --url "http://blog.craigabutler.com" --ignore-absolute-paths
> Fetching: http://ghost:2368/
> index.html                            100%[======================================================================>]  32.98K  --.-KB/s    in 0.005s
> robots.txt                            100%[======================================================================>]     107  --.-KB/s    in 0s
> assets/built/screen.css?v=2124e23fcd  100%[======================================================================>]  41.63K  --.-KB/s    in 0.005s
> public/cards.min.js?v=2124e23fcd      100%[======================================================================>]   7.23K  --.-KB/s    in 0s
> public/cards.min.css?v=2124e23fcd     100%[======================================================================>]  25.48K  --.-KB/s    in 0.004s
> test-post/index.html                  100%[======================================================================>]  13.74K  --.-KB/s    in 0.001s
> 
> [ ... ]
> 
> Fetching: http://ghost:2368/sitemap.xsl
> sitemap.xsl                           100%[======================================================================>]   6.39K  --.-KB/s    in 0s
> Fetching: http://ghost:2368/sitemap.xml
> sitemap.xml                           100%[======================================================================>]     638  --.-KB/s    in 0s
> Fetching: http://localhost:2368/sitemap-pages.xml
> ERROR: null
> Using Command: wget -q --show-progress --recursive --timestamping --page-requisites --no-parent --no-host-directories --restrict-file-name=unix --directory-prefix static --content-on-error  --trust-server-names http://localhost:2368/sitemap-pages.xml
> statichttp://localhost:2368/sitemap-pages.xml http://localhost:2368/sitemap-pages.xml Error: ENOENT: no such file or directory, open '/app/statichttp:/localhost:2368/sitemap-pages.xml'
>     at Object.openSync (fs.js:498:3)
>     at Object.readFileSync (fs.js:394:35)
>     at fetchUrlHelper (/usr/local/lib/node_modules/ghost-static-site-generator/src/helpers/fetchUrlHelper/fetchUrlHelper.js:47:31)
>     at Array.forEach (<anonymous>)
>     at /usr/local/lib/node_modules/ghost-static-site-generator/src/helpers/fetchUrlHelper/fetchUrlHelper.js:56:11
>     at Parser.<anonymous> (/usr/local/lib/node_modules/ghost-static-site-generator/node_modules/xml2js/lib/parser.js:304:18)
>     at Parser.emit (events.js:400:28)
>     at SAXParser.onclosetag (/usr/local/lib/node_modules/ghost-static-site-generator/node_modules/xml2js/lib/parser.js:262:26)
>     at emit (/usr/local/lib/node_modules/ghost-static-site-generator/node_modules/sax/lib/sax.js:624:35)
>     at emitNode (/usr/local/lib/node_modules/ghost-static-site-generator/node_modules/sax/lib/sax.js:629:5) {
>   errno: -2,
>   syscall: 'open',
>   code: 'ENOENT',
>   path: '/app/statichttp:/localhost:2368/sitemap-pages.xml'
> }
> ```
> 
> The affected files are:
> 
>     * sitemap-pages.xml
> 
>     * sitemap-posts.xml
> 
>     * sitemap-authors.xml
> 
>     * sitemap-tags.xml
> 
> 
> I haven't figured out what the problem is, but the fact that they're all XML files makes me think that it's some kind of issue in [replaceXmlUrlsHelper](https://github.com/Fried-Chicken/ghost-static-site-generator/tree/master/src/helpers/replaceXmlUrlsHelper) where `OPTIONS.SOURCE_DOMAIN_WITHOUT_PROTOCOL` is loading the default 'localhost' value instead of whatever was passed into the `--domain` option.
> 
> To address @imtomjorge's original post, looking at the commit history, I see that the `--sourceDomain` option was added in [2d8bfd8](https://github.com/Fried-Chicken/ghost-static-site-generator/commit/2d8bfd8ddd85846d14f9ab44e59d827e70cc151e), after the 1.1.4 release. In fact, GitHub thinks that the current version in the repository is 1.2.0, but the version provided by npm is still 1.1.4. @Fried-Chicken is there any reason you haven't pushed the newer version to npm?

have a look at #61, there's a couple of handy scripts there

**Comment by theyij** [comment]
Having the same issue here, tried the following command but did not work.
$ gssg ---domain "https://localhost:2368" --url "https://mysite.com" 

Does editing config.development.json file help?

**Comment by ItsIgnacioPortal** [comment]
I've fixed this bug in my fork of ghost-static-site-generator! https://github.com/itsignacioportal/ghost-static-site-generator/

**Comment by rija** [comment]
Hi @wunderhund,

I'm using Docker and I have met the same issue as you when running my ``gssg`` container service with 
```
command: /usr/local/bin/gssg --domain "http://ghost:2368" --dest "/var/www/public" --url $REMOTE_URL
```

I've almost fixed the problem by aliasing the ``ghost`` container service as ``localhost``.
This done by giving it a fixed IP address and then defining an ``extra_hosts`` section in the ``gssg`` container service:
```
extra_hosts:
    - "localhost:172.31.238.10"
```

The only problem left is the domain won't be replaced with the remote url for the Sitemap files. This was fixed by dropping ``--domain`` altogether and letting ``gssg`` thinks it's interacting with ``localhost`` for all files. After that, it all worked for me with the version of ``gssg`` available on NPM.  Below is the full ``docker-compose.yml`` file (I have a separate ``.env`` file to define the environment variables):

```
version: '3.7'

services:
  ghost:
    image: ghost:4.48.8-alpine
    volumes:
      - ./content:${GHOST_CONTENT}
      - ./config.production.json:${GHOST_INSTALL}/config.production.json
    ports:
      - ${GHOST_PORT}:2368
    networks:
      app-net:
        ipv4_address: 172.31.238.10

  gssg:
    build: .
    volumes:
      - "./static:/var/www"
    command: /usr/local/bin/gssg --dest "/var/www/public" --url $REMOTE_URL
    extra_hosts:
      - "localhost:172.31.238.10"
    networks:
      - app-net

networks:
  app-net:
    driver: bridge
    ipam:
      driver: default
      config:
        - subnet: 172.31.238.0/24
```

The ``Dockerfile`` used by the ``gssg`` container service is quite simple:

```
FROM node:19-alpine

RUN apk add --no-cache npm wget && \
	npm i -g ghost-static-site-generator
```

Usage:

```
$ docker-compose build gssg
$ docker-compose up -d ghost
$ docker-compose run --rm gssg
```

**Comment by Stunt0265** [comment]
Was this ever resolved?  Happening to me on Ubuntu 20.04

**Comment by ItsIgnacioPortal** [comment]
> Was this ever resolved? Happening to me on Ubuntu 20.04. Should I be able to install 1.2 using this command?
> 
> sudo npm install -g [Fried-Chicken/ghost-static-site-generator.git](https://github.com/Fried-Chicken/ghost-static-site-generator.git)

Look at #72
